Port 5/SECURE-TRANSPORT to vac research (#569)
vacp2p/rfc: content/docs/rfcs/53/README.md

---
slug: 53
title: 53/WAKU2-X3DH
name: X3DH usage for Waku payload encryption
status: stable
category: Standards Track
tags: waku-application
editor: Aaryamann Challani <aaryamann@status.im>
contributors:
- Andrea Piana <andreap@status.im>
- Pedro Pombeiro <pedro@status.im>
- Corey Petty <corey@status.im>
- Oskar Thorén <oskar@status.im>
- Dean Eigenmann <dean@status.im>
---

# Abstract

> This document describes a method can be used to provide a secure channel between two peers, and thus provide confidentiality, integrity, authenticity and forward secrecy. 

Reviewer: ```suggestion
This document describes a method that can be used to provide a secure channel between two peers, and thus provide confidentiality, integrity, authenticity and forward secrecy. 
```

Author: Addressed in b90e233

It is transport-agnostic and works over asynchronous networks.

It builds on the [X3DH](https://signal.org/docs/specifications/x3dh/) and [Double Ratchet](https://signal.org/docs/specifications/doubleratchet/) specifications, with some adaptations to operate in a decentralized environment.

# Background

This document describes how nodes establish a secure channel, and how various conversational security properties are achieved.

> # Specification

Reviewer: This caption should be removed. Seem to be copy&paste error :smile:

Author: hmm. I figured we needed motivation instead of background, does this work - [241a5f4](https://github.com/vacp2p/rfc/pull/569/commits/241a5f4fd9e14b30417668df8951a61e05e7ff40)?

Reviewer: Generally, Imo, we should stick more to spec parlance in RFC.
For the draft version OK though. Let's discuss and update this in a future editing pass before moving to stable.

> # Specification

Reviewer: ```suggestion
```

## Definitions

- **Perfect Forward Secrecy** is a feature of specific key-agreement protocols which provide assurances that session keys will not be compromised even if the private keys of the participants are compromised. 
Specifically, past messages cannot be decrypted by a third-party who manages to get a hold of a private key.

> - **Secret channel** describes a communication channel where Double Ratchet algorithm is in use.

Reviewer: ```suggestion
- **Secret channel** describes a communication channel where a Double Ratchet algorithm is in use.
```

Author: Addressed in b90e233

## Design Requirements

- **Confidentiality**: The adversary should not be able to learn what data is being exchanged between two Status clients.

> - **Authenticity**: The adversary should not be able to cause either endpoint of a Status 1:1 chat to accept data from any third party as though it came from the other endpoint.

Reviewer: With splitting the 1:1 chat spec, I'd remove `either endpoint of a Status 1:1 chat` and keep it more generic.

Author: Addressed in b90e233

- **Forward Secrecy**: The adversary should not be able to learn what data was exchanged between two Status clients if, at some later time, the adversary compromises one or both of the endpoint devices.
- **Integrity**: The adversary should not be able to cause either endpoint of a Status 1:1 chat to accept data that has been tampered with.

> 

Reviewer: Also, here I'd make it more generic. (see above)

Author: Addressed in b90e233

# Specification

## End-to-End Encryption

End-to-end encryption (E2EE) takes place between two clients. 
The main cryptographic protocol is a Double Ratchet protocol, which is derived from the [Off-the-Record protocol](https://otr.cypherpunks.ca/Protocol-v3-4.1.1.html), using a different ratchet. 
[Waku](https://rfc.vac.dev/spec/10/) subsequently encrypts the message payload, using symmetric key encryption. 
Furthermore, the concept of prekeys (through the use of [X3DH](https://signal.org/docs/specifications/x3dh/)) is used to allow the protocol to operate in an asynchronous environment.
It is not necessary for two parties to be online at the same time to initiate an encrypted conversation.

## Cryptographic Protocols

This protocol uses the following cryptographic primitives:
- X3DH
    - Elliptic curve Diffie-Hellman key exchange (secp256k1)
    - KECCAK-256
    - ECDSA
    - ECIES
- Double Ratchet
    - HMAC-SHA-256 as MAC
    - Elliptic curve Diffie-Hellman key exchange (Curve25519)
    - AES-256-CTR with HMAC-SHA-256 and IV derived alongside an encryption key

    The node achieves key derivation using HKDF.

> 

Reviewer: I'd link the RFC: https://www.rfc-editor.org/rfc/rfc5869

Author: Addressed in b90e233

Since bundles stored in QR codes or ENS records cannot be updated to delete already used keys, the bundle MAY be rotated every 24 hours, and distributed via Waku.

## Flow

The key exchange can be summarized as follows:

1. Initial key exchange: Two parties, Alice and Bob, exchange their prekey bundles, and derive a shared secret.

2. Double Ratchet: The two parties use the shared secret to derive a new encryption key for each message they send.

3. Chain key update: The two parties update their chain keys. The chain key is used to derive new encryption keys for future messages.

4. Message key derivation: The two parties derive a new message key from their chain key, and use it to encrypt a message.

### 1. Initial key exchange flow (X3DH)

[Section 3 of the X3DH protocol](https://signal.org/docs/specifications/x3dh/#sending-the-initial-message) describes the initial key exchange flow, with some additional context:
- The peers' identity keys `IK_A` and `IK_B` correspond to their public keys;
- Since it is not possible to guarantee that a prekey will be used only once in a decentralized world, the one-time prekey `OPK_B` is not used in this scenario;

> - Nodes SHOULD not send Bundles to a centralized server, but instead served in a decentralized way as described in the [Pre-keys section](#pre-keys).

Reviewer: ```suggestion
- Nodes SHOULD not send Bundles to a centralized server, but instead provide them in a decentralized way as described in the [Pre-keys section](#pre-keys).
```

Author: Addressed in b90e233

---
slug: 53
title: 53/WAKU2-X3DH
name: X3DH usage for Waku payload encryption

> status: raw

Reviewer 2: draft? or stable even

Author: It was labelled stable initially, since it is in use by the users of the Status app. as per https://github.com/vacp2p/rfc/pull/569#pullrequestreview-1281254759, was moved to `raw` to indicate lifecycle through vac research. Happy to change it to `draft` though

Reviewer: I'd suggest draft.
